File: leetcode/stack/739. Daily Temperatures/py/main.py

class Solution:
    # Time O(N)
    # Space O(N)
    def dailyTemperatures(self, T):
        st, N = [], len(T)
        ans = [0] * N

        for i in reversed(range(N)):
          while st and T[st[-1]] <= T[i]:
            st.pop()

          if st:
            ans[i] = st[-1] - i
          
          st.append(i)

        return ans

    # Time O(N * W)
    # Space O(N + W)
    def dailyTemperatures_II(self, T):
        n, inf = len(T), float('inf')
        ans, temp_indexes= [0] * n, [inf] * 101

        for i in reversed(range(n)):
          warmer_index = inf

          for j in range(T[i] + 1, 101):
            # find min index
            warmer_index = min(warmer_index, temp_indexes[j])    
            
          if warmer_index < inf:
            ans[i] = warmer_index - i

          # store index
          temp_indexes[T[i]] = i

        return ans

File: leetcode/stack/739. Daily Temperatures/js/main.js
// Time O(N)
// Space O(N)
const dailyTemperatures = T => {
  let st = [];
  let n = T.length;
  let res = [];

  for (let i = 0; i < n; i++) {
    while (st.length && T[st[st.length - 1]] < T[i]) {
      let j = st.pop();
      res[j] = i - j;
    }

    st.push(i);
  }

  while (st.length) {
    res[st.pop()] = 0;
  }

  return res;
};
